lib/puppetserver/ca/action/prune.rb

```diff
@@ -0,0 +1,134 @@
+require 'optparse'
+require 'openssl'
+require 'puppetserver/ca/errors'
+require 'puppetserver/ca/utils/cli_parsing'
+require 'puppetserver/ca/utils/file_system'
+require 'puppetserver/ca/utils/config'
+require 'puppetserver/ca/x509_loader'
+
+module Puppetserver
+  module Ca
+    module Action
+      class Prune
+        include Puppetserver::Ca::Utils
+
+        SUMMARY = "Prune the local CRL on disk to remove any duplicated certificates"
+        BANNER = <<-BANNER
+Usage:
+  puppetserver ca prune [--help]
+  puppetserver ca prune [--config]
+
+Description:
+  Prune the list of revoked certificates of any duplication within it.  This command
+  will only prune the CRL issued by Puppet's CA cert.
+
+Options:
+BANNER
+
+        def initialize(logger)
+          @logger = logger
+        end
+
+        def run(inputs)
+          config_path = inputs['config']
+
+          # Validate the config path.
+          if config_path
+            errors = FileSystem.validate_file_paths(config_path)
+            return 1 if Errors.handle_with_usage(@logger, errors)
+          end
+
+          # Validate puppet config setting.
+          puppet = Config::Puppet.new(config_path)
+          puppet.load(logger: @logger)
+          return 1 if Errors.handle_with_usage(@logger, puppet.errors)
+
+          # Validate that we are offline
+          return 1 if HttpClient.check_server_online(puppet.settings, @logger)
+
+          # Getting the CRL(s)
+          loader = X509Loader.new(puppet.settings[:cacert], puppet.settings[:cakey], puppet.settings[:cacrl])
+
+          # Take 'em
+          puppet_crl = loader.crls.select { |crl| crl.verify(loader.key) }
+          # Prune 'em
+          prune_CRLs(puppet_crl)
+          # Update 'em
+          update_pruned_CRL(puppet_crl, loader.key)
+          # Write 'em out
+          write_crl_to_file(loader.crls, puppet.settings[:cacrl])
+
+          @logger.inform("Finished pruning Puppet's CRL")
+          return 0
+        end
+
+        def prune_CRLs(crl_list)
+          crl_list.each do |crl|
+            existed_serial_number = Set.new()
+            revoked_list = crl.revoked
+            @logger.debug("Pruning duplicate entries in CRL for issuer " \
+              "#{crl.issuer.to_s(OpenSSL::X509::Name::RFC2253)}") if @logger.debug?
+
+            revoked_list.delete_if do |revoked|
+              if existed_serial_number.add?(revoked.serial)
+                false
+              else
+                @logger.debug("Remove duplicate of #{revoked.serial}, " \
+                  "revoked on #{revoked.time}\n") if @logger.debug?
```

Comment:
Found out that calling the getter function for 150k times will indeed slow down runtime.  Thus I reimplemented a check for verbose mode status.  That way we don't need to call the getter functions all the time and get rejected because we're not in verbose mode.  The difference in run time was from 0.5 - 1 second.  It's not a huge impact on time, but it's noticeable.

Reply:
It's the `time` and `serial` function.  They are essentially getter functions for the X509::Revoked class.

Reply:
I think the time is valuable to have, so you could get an understanding of when duplicates are appearing.

Reply:
fair